flow/data_pipeline/data_pipeline.py

```diff
@@ -0,0 +1,284 @@
+    extra_info: dict
+        extra information needed in the trajectory table, collected from flow
+    partition_name: str
+        the name of the partition to put this output to
+    Returns
+    -------
+    output_file_path: str
+        the local path of the outputted csv file that should be used for
+        upload to s3 only, it does not the human readable column names and
+        will be deleted after uploading to s3. A copy of this file with all
+        the column name will remain in the ./data folder
+    """
+    extra_info = pd.DataFrame.from_dict(extra_info)
+    # extra_info["partition"] = partition_name
+    extra_info.to_csv(data_path, index=False)
+    upload_only_file_path = data_path[:-4] + "_upload" + ".csv"
+    extra_info.to_csv(upload_only_file_path, index=False, header=False)
+    return upload_only_file_path
+
+
+def upload_to_s3(bucket_name, bucket_key, file_path, only_query):
+    """Upload a file to S3 bucket.
+
+    Parameters
+    ----------
+    bucket_name : str
+        the bucket to upload to
+    bucket_key: str
+        the key within the bucket for the file
+    file_path: str
+        the path of the file to be uploaded
+    only_query: str
+        specify which query should be run on this file by lambda:
+        if empty: run none of them
+        if "all": run all available analysis query
+        if a string of list of queries: run only those mentioned in the list
+    """
+    s3 = boto3.resource("s3")
+    s3.Bucket(bucket_name).upload_file(file_path, bucket_key,
+                                       ExtraArgs={"Metadata": {"run-query": only_query}})
+    return
+
+
+class AthenaQuery:
+    """
+    Class used to run query.
+
+    Act as a query engine, maintains an open session with AWS Athena.
+
+    Attributes
+    ----------
+    MAX_WAIT: int
+        maximum number of seconds to wait before declares time-out
+    client: boto3.client
+        the athena client that is used to run the query
+    existing_partitions: list
+        a list of partitions that is already recorded in Athena's datalog,
+        this is obtained through query at the initialization of this class
+        instance.
+    """
+
+    def __init__(self):
+        """Initialize AthenaQuery instance.
+
+        initialize a client session with AWS Athena,
+        query Athena to obtain extisting_partition.
+        """
+        self.MAX_WAIT = 60
+        self.client = boto3.client("athena")
+        self.existing_partitions = self.get_existing_partitions()
+
+    def get_existing_partitions(self):
+        """Return the existing partitions in the S3 bucket.
+
+        Returns
+        -------
+        partitions: a list of existing partitions on S3 bucket
+        """
+        response = self.client.start_query_execution(
+            QueryString='SHOW PARTITIONS trajectory_table',
+            QueryExecutionContext={
+                'Database': 'simulation'
+            },
+            WorkGroup='primary'
```

Comment:
What exactly is the workgroup?

Reply:
A workgroup is used in Athena to control access policy and costs. Each workgroup has separate history and configurations. An account must run query from an existing workgroup. The `primary` workgroup is the default one. We can create a new one later if needed.